include/nbl/system/demote_promote_writer_readers_lock.h

```diff
@@ -130,62 +163,75 @@ class demote_promote_writer_readers_lock
+			assert(!(oldState & currentReadersMask));
+
 			// release the flipLock, decrement currentReaders
-			state.store(actual);
+			state_lock_value_t newState = oldState & ~writingMask;
+			state.store(newState);
 			break;
 		}
 	}
 
 	/**
-	* @brief Upgrades a thread with a reader's lock to a writer's lock. If a thread already has a reader's lock, use this instead of `read_unlock()` -> `write_lock()` 
+	* @brief Upgrades a thread with a reader's lock to a writer's lock. If a thread already has a reader's lock, use this instead of `read_unlock()` -> `write_lock()`
+	*        This thread will be blocked until there are no more readers and there's no thread currently writing. It will also preempt new readers and writers from 
+	*        acquiring the lock.
 	*/
 	void upgrade()
 	{
+		constexpr state_lock_value_t preemptedMask = currentReadersMask;
 		bool registeredPending = false;
-		for (uint32_t spinCount = 0; true; spinCount++)
+		for (state_lock_value_t spinCount = 0; true; spinCount++)
 		{
-			uint32_t actual = state.fetch_or(flipLock);
+			const state_lock_value_t oldState = state.fetch_or(flipLock);
 			// stateLocked: some other thread is doing work and expects to update state
-			if (actual & flipLock)
+			if (oldState & flipLock)
 			{
 				if (spinCount > SpinsBeforeYield)
 					std::this_thread::yield();
 				continue;
 			}
-			// If no one's writing, take lock and release the flipLock
-			if (!(actual & writingMask))
+
+			// Sanity check: ensure no one's writing if we had a reader's lock
+			assert(!(oldState & writingMask));
+			// Sanity check: if this thread is reading, then `currentReaders` can't be 0
+			assert(oldState & currentReadersMask);
+
+			state_lock_value_t nextState = oldState;
+			// If there are threads currently reading or writing, must release flipLock for other threads to progress
+			if (nextState & preemptedMask)
 			{
-				actual |= writingMask;
-				// If had been registered earlier
-				if (registeredPending)
+				// Register pending if not done so earlier
+				if (!registeredPending)
 				{
-					actual -= static_cast<uint32_t>(StateSemantics{ .pendingUpgrades = 1 });
+					registeredPending = true;
+					nextState += impl::DPWRLStateSemantics{ .currentReaders = 0, .pendingWriters = 0, .pendingUpgrades = 1, .writing = false, .stateLocked = false };
 				}
-				state.store(actual);
-				break;
-			}
-			// Register pending if not done so earlier
-			if (!registeredPending)
-			{
-				registeredPending = true;
-				actual += static_cast<uint32_t>(StateSemantics{ .pendingUpgrades = 1 });
-			}
-			// Release flipLock
-			state.store(actual);
-			if (spinCount > SpinsBeforeYield)
-				std::this_thread::yield();
+				// Release flipLock
+				state.store(nextState);
+				if (spinCount > SpinsBeforeYield)
+					std::this_thread::yield();
 			}
+			// Release flipLock, declare that this thread is writing, mark that this thread is no longer pending IF had been registered as such
```

Comment:
you're missing a `continue` after the yield/spin

This is a great argument for keeping your code DRY and unifying the control flow 
https://github.com/Devsh-Graphics-Programming/Nabla/pull/845#discussion_r1986890925